### PR TITLE
Remove AutoFormat

### PR DESCRIPTION
**Pull request by philei-tt**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/17276

### Problem description
run_with/without_autoformat are strange functions that are almost not used anywhere. Also in practice they usually just convert inputs to tile layout before calling op. 

### What's changed
Remove: 
- [x] run_with_autoformat 
- [x] run_without_autoformat
- [x] The whole AutoFormat class ( move_tensor_to_mem_config, move_tensor_to_mem_config, format_output_tensor, format_input_tensor, check_input_tensor_format. move_tensor_to_device_and_pad),
- [X] Usage and python bindings for format_input/output_tensor

Other:
- Remove usage and python bindings for format_input/output_tensor
- Move AutoFormat::Get/SetDefaultDevice to ttnn/core/device.cpp
- Move AutoFormat::pad_to_tile_shape to ttnn/operations/data_movement/common/common.cpp

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes -- https://github.com/tenstorrent/tt-metal/actions/runs/19638796567 
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable) -- https://github.com/tenstorrent/tt-metal/actions/runs/19638799983 
- [x] New/Existing tests provide coverage for changes
- [x] (Single-card) Model perf tests -- https://github.com/tenstorrent/tt-metal/actions/runs/19669180469
- [x] (Single-card) Device perf regressions -- https://github.com/tenstorrent/tt-metal/actions/runs/19669190209

### CODE REVIEW

**Review by ayerofieiev-tt — COMMENTED** [review]
It is pretty funny that this PR removed more code than adds.
Something that was supposed to reduce complexity by being put in infra level just made things worse..

Thank you!

**Comment by philei-tt** [comment]
@ayerofieiev-tt There are some utilities from auto_format.cpp that are used in many places and some of them are even pybinded. It would take a lot more time to remove them + I'm afraid to break python api. I suggest addressing that later. 
- GetDefaultDevice
- SetDefaultDevice
- pad_to_tile_shape
- check_input_tensor_format
- move_tensor_to_device_and_pad
- format_input_tensor
- format_output_tensor

File: ttnn/cpp/ttnn/operations/data_movement/bcast/bcast.cpp [hunk 1]
```diff
@@ -78,12 +78,36 @@ Tensor BcastOperation::invoke(
...
+    Tensor formatted_b = format_input(input_tensor_b);
+
+    return tt::tt_metal::operation::run(
                EltwiseBinaryBroadcast{bcast_op, bcast_dim, output_memory_config},
-               {input_tensor_a, input_tensor_b},
+               {formatted_a, formatted_b},
                {},
-               {output_tensor},
-               0 /* pad_value*/)
+               {output_tensor})
```
[the remark this answers is not in the document]
Comment (philei-tt) [reply ttnn/cpp/ttnn/operations/data_movement/bcast/bcast.cpp]: ???

**Review by bbradelTT — APPROVED** [review]
If you expect any perf changes you should run more pipelines.

**Review by sjameelTT — APPROVED** [review]

**Review by dstoiljkovicTT — APPROVED** [review]

**Review by dchenTT — APPROVED** [review]

**Review by bbradelTT — CHANGES_REQUESTED** [review]
Can you please test with inputs being row major?

The following seems to fail in this branch and pass on main:
```
from loguru import logger
import random
import pytest
import torch
import ttnn
import traceback

from tests.ttnn.utils_for_testing import assert_with_pcc, assert_allclose
@pytest.mark.parametrize(
        "input_shape, dim, keepdim", [
((512, 1024, 1, 2), -1, False),
            ]
        )
def test_mean(device, input_shape, dim, keepdim):
    torch.manual_seed(0)
    torch_input_tensor = torch.rand(input_shape, dtype=torch.float32)
    torch_output_tensor = torch.mean(torch_input_tensor, dim, keepdim)
        
    input_tensor = ttnn.from_torch(torch_input_tensor, dtype=ttnn.float32,
            #layout=ttnn.TILE_LAYOUT,
            device=device)
    
    output_tensor = ttnn.mean(input_tensor, dim=dim, keepdim=keepdim)
    output_tensor = ttnn.to_torch(output_tensor)
            
    assert_with_pcc(torch_output_tensor, output_tensor, pcc=0.999)
```

File: ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.hpp [hunk 2]
```diff
@@ -96,6 +96,20 @@ class AutoFormat {
...
 
+    /**
+     * Formats an input tensor a given layout + memory_config. If padding is needed it uses pad_value (e.g. RM->TILE)
+     * @param input Input tensor
+     * @param pad_value Value to use for padding
+     * @param target_layout Desired layout
+     * @param target_mem_config Optional memory configuration
+     * @return Formatted tensor
+     */
+    static Tensor format_input_tensor(
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.hpp]: does it have to be "input" tensor? :)

File: ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding/rotary_embedding.cpp [hunk 3]
```diff
@@ -71,25 +71,10 @@ ttnn::Tensor RotaryEmbeddingOperation::invoke(
...
-            return ttnn::tilize_with_val_padding(input, tile_aligned_shape, pad_val, input.memory_config());
-        }
-    };
-
-    Tensor formatted_input = format_input(input_tensor);
-    Tensor formatted_cos = format_input(cos_cache);
-    Tensor formatted_sin = format_input(sin_cache);
+    using namespace ttnn::operations::experimental::auto_format;
+    Tensor formatted_input = AutoFormat::format_input_tensor(input_tensor, 0, Layout::TILE);
+    Tensor formatted_cos = AutoFormat::format_input_tensor(cos_cache, 0, Layout::TILE);
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding/rotary_embedding.cpp]: no idea what 0 is ;)

File: ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.hpp [hunk 4]
```diff
@@ -96,6 +96,20 @@ class AutoFormat {
...
+        const Tensor& input,
+        float pad_value,
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.hpp]: pad value makes no sense for row major

File: ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.hpp [hunk 5]
```diff
@@ -96,6 +96,20 @@ class AutoFormat {
...
+        tt::tt_metal::Layout target_layout,
+        std::optional<tt::tt_metal::MemoryConfig> target_mem_config = std::nullopt);
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.hpp]: is this only to say SRAM / DRAM or also for sharding?

File: ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding/rotary_embedding.cpp (hunk 3 above)
Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding/rotary_embedding.cpp]: why is this not smth like "tilize(tensor, PadValue(value))" ?

File: ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.hpp (hunk 5 above)
Comment (philei-tt) [reply ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.hpp]: This is just passed to tilize/pad/untilize/tilize_with_val_padding inside format_(input)tensor. So this is basically output memory config, can be anything as long as op supports that

File: ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding/rotary_embedding.cpp (hunk 3 above)
Comment (philei-tt) [reply ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding/rotary_embedding.cpp]: Using explicit PadValue is a good idea for readability, added

Comment (philei-tt) [reply ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding/rotary_embedding.cpp]: But this is not just tilize. It basically converts to target layout + memory config (+uses given padding if needs to tilize). Not sure what's better name.

File: ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.hpp (hunk 2 above)
Comment (philei-tt) [reply ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.hpp]: Renamed to format_tensor

File: ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.hpp (hunk 4 above)
Comment (philei-tt) [reply ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.hpp]: It actually does. When input is RM you can still pass padded_shape > logical_shape, and then it will use ttnn::pad with a given value.

**Review by cglagovichTT — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/reduce/test_row_major_reduce.py [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+
+#     # Create tensor without specifying layout - defaults to ROW_MAJOR
+#     input_tensor = ttnn.from_torch(torch_input_tensor, dtype=ttnn.bfloat16, device=device)
+
+#     assert input_tensor.layout == ttnn.ROW_MAJOR_LAYOUT, "Input should be in ROW_MAJOR_LAYOUT"
+
+#     output_tensor = ttnn.sum(input_tensor)
+#     output_tensor = ttnn.to_torch(output_tensor)
+
+#     assert_with_pcc(torch_output_tensor, output_tensor, pcc=0.999)
```
[the remark this answers is not in the document]
Comment (philei-tt) [reply tests/ttnn/unit_tests/operations/reduce/test_row_major_reduce.py]: Good catch

[the remark this answers is not in the document]
Comment (philei-tt) [reply tests/ttnn/unit_tests/operations/reduce/test_row_major_reduce.py]: Fixed

**Review by bbradelTT — APPROVED** [review]

**Review by aczajkowskiTT — APPROVED** [review]

File: ttnn/cpp/ttnn-pybind/device.cpp [hunk 7]
```diff
@@ -289,10 +289,7 @@ void device_module(py::module& m_device) {
 
     m_device.def(
         "GetDefaultDevice",
-        []() {
-            return dynamic_cast<MeshDevice*>(
-                ttnn::operations::experimental::auto_format::AutoFormat::GetDefaultDevice());
-        },
+        []() { return dynamic_cast<MeshDevice*>(ttnn::GetDefaultDevice()); },
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn-pybind/device.cpp]: Dynamic cast not needed right?

File: ttnn/cpp/ttnn/operations/data_movement/bcast/bcast.cpp [hunk 8]
```diff
@@ -78,13 +78,22 @@ Tensor BcastOperation::invoke(
...
+    Tensor formatted_b = ttnn::tilize_with_val_padding(
+        input_tensor_b, padded_shape_b, tt::tt_metal::PadValue(0.0f), input_tensor_b.memory_config());
+
+    auto output = tt::tt_metal::operation::run(
+                      EltwiseBinaryBroadcast{bcast_op, bcast_dim, output_memory_config},
+                      {input_tensor_a, input_tensor_b},
+                      {},
+                      {output_tensor})
+                      .at(0);
+    return output;
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/bcast/bcast.cpp:96]: I thought run with auto format does
* change layout of tensors a/b 
* runs the op 
* restores the layout

Not the case here?

File: ttnn/cpp/ttnn/operations/data_movement/common/common.cpp [hunk 9]
```diff
@@ -575,6 +575,26 @@ ttnn::Shape compute_padded_shape(ttnn::Shape logical_shape, const uint32_t tile_
     return ttnn::Shape(output_shape_vec);
 }
 
+ttnn::Shape pad_to_tile_shape(const ttnn::Shape& unpadded_shape) {
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/common/common.cpp:578]: A bit of a duplication of TensorSpec/TensorLayout functionality 

Keep in mind that tile size is also not constant

**Review by ayerofieiev-tt — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/bcast/bcast.cpp (hunk 8 above)
Comment (philei-tt) [reply ttnn/cpp/ttnn/operations/data_movement/bcast/bcast.cpp:96]: I thought so too. This is stupid, but it was not restoring layout, there was hardcoded tile layout: https://github.com/tenstorrent/tt-metal/blob/63bb00ab4af2ec88203bbb371291324b7d2a4af1/ttnn/core/run_operation.cpp#L216

File: ttnn/cpp/ttnn/operations/data_movement/common/common.cpp (hunk 9 above)
Comment (philei-tt) [reply ttnn/cpp/ttnn/operations/data_movement/common/common.cpp:578]: I know, unfortunately this function is pybinded, and used in many tests and models, so it's a bit messy to remove it. PR is big enough

File: ttnn/cpp/ttnn-pybind/device.cpp (hunk 7 above)
Comment (philei-tt) [reply ttnn/cpp/ttnn-pybind/device.cpp]: Will remove

**Review by gwangTT — APPROVED** [review]
